From a88fb8ae5c52fffdca5038cbe3261d52bc92101c Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Tue, 18 Nov 2025 15:04:45 -0500
Subject: [PATCH] Issue 7007 - Improve paged result search locking

Description:

Hold the paged result connection hash mutex while acquiring the global
connection paged result lock. Otherwise there is a window where the
mutex could be rmoved and lead to a crash

Relates: https://github.com/389ds/389-ds-base/issues/7007

Reviewed by: progier, spichugi, and tbordaz(Thanks!!!)
---
 ldap/servers/slapd/pagedresults.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/ldap/servers/slapd/pagedresults.c b/ldap/servers/slapd/pagedresults.c
index c3f3aae01..941ab97e3 100644
--- a/ldap/servers/slapd/pagedresults.c
+++ b/ldap/servers/slapd/pagedresults.c
@@ -804,6 +804,7 @@ pagedresults_cleanup(Connection *conn, int needlock)
         prp->pr_current_be = NULL;
         if (prp->pr_mutex) {
             PR_DestroyLock(prp->pr_mutex);
+            prp->pr_mutex = NULL;
         }
         memset(prp, '\0', sizeof(PagedResults));
     }
@@ -841,6 +842,7 @@ pagedresults_cleanup_all(Connection *conn, int needlock)
         prp = conn->c_pagedresults.prl_list + i;
         if (prp->pr_mutex) {
             PR_DestroyLock(prp->pr_mutex);
+            prp->pr_mutex = NULL;
         }
         if (prp->pr_current_be && prp->pr_search_result_set &&
             prp->pr_current_be->be_search_results_release) {
@@ -1022,11 +1024,10 @@ pagedresults_lock(Connection *conn, int index)
     }
     pthread_mutex_lock(pageresult_lock_get_addr(conn));
     prp = conn->c_pagedresults.prl_list + index;
-    pthread_mutex_unlock(pageresult_lock_get_addr(conn));
     if (prp->pr_mutex) {
         PR_Lock(prp->pr_mutex);
     }
-    return;
+    pthread_mutex_unlock(pageresult_lock_get_addr(conn));
 }
 
 void
@@ -1038,11 +1039,10 @@ pagedresults_unlock(Connection *conn, int index)
     }
     pthread_mutex_lock(pageresult_lock_get_addr(conn));
     prp = conn->c_pagedresults.prl_list + index;
-    pthread_mutex_unlock(pageresult_lock_get_addr(conn));
     if (prp->pr_mutex) {
         PR_Unlock(prp->pr_mutex);
     }
-    return;
+    pthread_mutex_unlock(pageresult_lock_get_addr(conn));
 }
 
 int
-- 
2.52.0

